pkg/apihelpers/apihelpers.go

```diff
@@ -324,6 +325,43 @@ func IsControllerConfigCompleted(ccName string, ccGetter func(string) (*mcfgv1.C
 	return fmt.Errorf("ControllerConfig has not completed: completed(%v) running(%v) failing(%v)", completed, running, failing)
 }
 
+// AreMCGeneratingSubControllersCompleted checks whether all MC producing sub-controllers are completed
+func AreMCGeneratingSubControllersCompleted(crcLister func(labels.Selector) ([]*mcfgv1.ContainerRuntimeConfig, error), mckLister func(labels.Selector) ([]*mcfgv1.KubeletConfig, error), selector labels.Selector) error {
```

Comment:
This function covers kubelet and containerruntime controller generation, which covers quite a bit of the sub-controller generation, but I'm curious if this actually is sufficient for the bug in question, since it originally referred to a `node.config.openshift.io` object, which is another thing the kubelet subcontrollers sync on (just like how the containerruntime can also sync off of image.config objects)

The other scenario to think through is for the additional MCs we generate to set defaults, etc. that are not directly attached to a kubelet or containerruntime config object, whether this would work (although I guess we don't really need to be future proof here, if we do add more sub-controller-generated-MCs, it should also be the responsibility of the PR to update this?)

Now I think about it, maybe my concerns are not applicable here, because the two main syncs here (containerruntime and kubelet) finishing would imply sub-controllers have succeeded, and if it was going to generate something off of something other than kubelet/containerruntime, it should have done so already in that sync loop, so the finishing of everything else is implicit. If that's the case, I'm fine with what we have as well.

(maybe to be safe, add in checks for node.config syncs? I don't remember if they also update observedgeneration or not)

Reply:
https://github.com/openshift/machine-config-operator/blob/657b65f9cf4df675ea50cbf9e0cf30a04f131f6f/pkg/controller/kubelet-config/kubelet_config_nodes.go#L179 
This suggests to me that whenever a node config changes, it automatically triggers a sync to all the existing KubeletConfigs. Hence the kubelet configs are modified as well. Thus, by just watching the status of the current KubeletConfigs aimed at the MCP, we should be able to determine if they were successful.

However, what this also means is if we fail to apply a valid Kubelet Config or a ContainerRuntime config due to any reason, no new rendered-MC will be produced till the render-controller syncs again

Reply:
Hmm, ok, if what you described is correct I think the behaviour should be fine as well. Node.config aside, I'm wondering if there are other sync loops we need to guard for, but maybe we're all covered through the main syncs in kubelet and containerruntime